src/blockdata/transaction.rs

```diff
@@ -559,8 +560,113 @@ impl Transaction {
 
         Ok(amount)
     }
+
+    /// Compute the message to be signed by the CLSAG signature algorithm.
+    ///
+    /// The message consists of three parts:
+    ///
+    /// 1. The hash of the transaction prefix.
+    /// 2. The hash of a consensus-encoded [`RctSigBase`].
+    /// 3. The hash of all bulletproofs.
+    pub fn signature_hash(&self) -> Result<hash::Hash, SignatureHashError> {
```

Comment:
I agree that it is quite difficult to unit test this. I'm fine merging it in main but we add

```rust
#[cfg(feature = "experimental")]
#[cfg_attr(docsrs, doc(cfg(feature = "experimental")))]
```
so it is not compiled in by default and the feature name warns you that this may be incorrect/not fully tested.

It also seems that the function only works for CLSAG hashes, as mentioned in the doc comment, but the function name is generic. I propose to change the function name or check the `RctType` and return an error for unsupported types.

Reply:
Happy to feature-flag this with `experimental`.

> It also seems that the function only works for CLSAG hashes, as mentioned in the doc comment, but the function name is generic. I propose to change the function name or check the RctType and return an error for unsupported types.

What do you think about leaving the function name generic and fail for now if we have an unsupported `RctType`? We can always add support for more `RctType`s within the same function so that the function name makes sense. I don't think different public(!) functions for the individual `RctType`s make sense. We might want to have private ones for clarity though.

Reply:
Yes, I agree that only one public function should exist.